<a href="https://colab.research.google.com/github/Dieljr/README/blob/main/SVR_Grid.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# Conteúdo para aulas online
## Terceiro Notebook: GridSearchCV para SVM

### O Objetivo desse notebook é o de apresentar recursos intermediários a avançados sobre o Algoritmo SUPPORT VECTOR MACHINE REGRESSOR com GridSearchCV. O SVR é da classe dos SVMs.

1. Source: 

a) <font color = '#CD5C5C'>Site do sklearn:</font> https://scikit-learn.org/stable/modules/generated/sklearn.model_selection.GridSearchCV.html?highlight=grid#sklearn.model_selection.GridSearchCV

b) <font color = '#CD5C5C'>Site do Sklearn: SVR:</font>https://scikit-learn.org/stable/modules/generated/sklearn.svm.LinearSVR.html#sklearn.svm.LinearSVR

b) <font colo = '#CD5C5C'>Livro Mãos à Obra: Aprendizado de Máquina com Scikit-Learn e TensorFlow</font>: https://www.amazon.com.br/M%C3%A3os-obra-aprendizado-Scikit-Learn-inteligentes/dp/8550815489/ref=sr_1_1?crid=2JG8SYAU8K3UQ&keywords=tensorflow&qid=1646567493&s=books&sprefix=tensort+flow%2Cstripbooks%2C229&sr=1-1&ufe=app_do%3Aamzn1.fos.6d798eae-cadf-45de-946a-f477d47705b9

c) <font color = '#CD5C5C'>Livro Análise de dados com Python e Pandas</font>:https://www.amazon.com.br/An%C3%A1lise-Dados-com-Python-Pandas/dp/8575226991/ref=sr_1_1?__mk_pt_BR=%C3%85M%C3%85%C5%BD%C3%95%C3%91&keywords=analise+de+dados+com+python&qid=1646671469&sr=8-1&ufe=app_do%3Aamzn1.fos.6d798eae-cadf-45de-946a-f477d47705b9


3. Parametros ou hiperparamentros:
Para você que está começando, pode parecer estranho e confuso, mas é assim que a comunidade em Ciência de Dados se refere ajustadores dos algoritmos de machine learning. Então, vá se acostumando!
Irei apresentar somente os mais importantes parametros que o ajudará no seu dia-a-dia como cientista de dados.

class sklearn.svm.LinearSVR(*, epsilon=0.0, tol=0.0001, C=1.0, loss='epsilon_insensitive', fit_intercept=True, intercept_scaling=1.0, dual=True, verbose=0, random_state=None, max_iter=1000)

a) epsilon: 0.0 por padrão. Quanto menor seu valor, menor as margens entre as duas classes.

b) C: 1.0 por padrão. Quanto menor seu valor, maior as margens entre as duas classes. É chamado de parêmetro de regularização. Para uma grande número de dados e aconselhável reduzir o valor de C.

c) loss: métodos de funções de perda. É ‘epsilon_insensitive’ por padrão, mas pode ser alterado para ‘squared_epsilon_insensitive’.  O epsilon-insensitive significa a capacidade de inserir mais instâncias dentro da margem e não afetar as previsões do modelo. Esse parametro controla a largura da banda entre as classes.

d) verbose: print as saídas do algoritmo. É 0 por padrão.

Atributos: ou o que se pode extrair do algoritmo depois que ele for usado. Melhor do que explicá-los aqui (teoria), é apresentá-los na prática!

Cometários:
Uma boa combinação é o uso dos parâmetros C e epsilon para capturar a melhor configuração das margens para REGRESSÃO. Para a versão LinearSVC o objetivo é o inverso.
LinearSVR utiliza o kernel 'liblinear'.
O algotrimo recomenda que você  redimensione os dados X para [0,1] ou [-1,+1] ou padronize par média = 0 e desvio padrão = 1.

## PRÁTICA:

In [ ]:
# Carregar pacotes

import numpy as np
import pandas as pd

import matplotlib.pyplot as plt
import matplotlib as mpl
import seaborn as sb

from sklearn.preprocessing import StandardScaler
from sklearn.svm import LinearSVR
from sklearn.model_selection import train_test_split, GridSearchCV

import sys

In [ ]:
print(sys.version)

3.7.6 (default, Jan  8 2020, 20:23:39) [MSC v.1916 64 bit (AMD64)]


In [ ]:
# Carregar dataset
df = pd.read_csv('mpg.csv', sep = ';', names=['mpg', 'cylinders', 'displacement', 'horsepower', 'weight', 'accelaration',
                                              'model_year', 'origin', 'car_name'])
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 406 entries, 0 to 405
Data columns (total 9 columns):
 #   Column        Non-Null Count  Dtype  
---  ------        --------------  -----  
 0   mpg           398 non-null    float64
 1   cylinders     406 non-null    float64
 2   displacement  406 non-null    float64
 3   horsepower    400 non-null    float64
 4   weight        406 non-null    float64
 5   accelaration  406 non-null    float64
 6   model_year    406 non-null    float64
 7   origin        1 non-null      object 
 8   car_name      0 non-null      float64
dtypes: float64(8), object(1)
memory usage: 28.7+ KB


### Pequenos ajustes

In [ ]:
# Não precisaremos das duas últimas colunas
del df['origin']
del df['car_name']

In [ ]:
df.head()

mpg  cylinders  displacement  horsepower  weight  accelaration  model_year
0  18.0        8.0         307.0       130.0  3504.0          12.0        70.0
1  15.0        8.0         350.0       165.0  3693.0          11.5        70.0
2  18.0        8.0         318.0       150.0  3436.0          11.0        70.0
3  16.0        8.0         304.0       150.0  3433.0          12.0        70.0
4  17.0        8.0         302.0       140.0  3449.0          10.5        70.0

In [ ]:
# Podemos excluir os valores missing sem perda significativa.
df = df.dropna()

In [ ]:
df.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 392 entries, 0 to 405
Data columns (total 7 columns):
 #   Column        Non-Null Count  Dtype  
---  ------        --------------  -----  
 0   mpg           392 non-null    float64
 1   cylinders     392 non-null    float64
 2   displacement  392 non-null    float64
 3   horsepower    392 non-null    float64
 4   weight        392 non-null    float64
 5   accelaration  392 non-null    float64
 6   model_year    392 non-null    float64
dtypes: float64(7)
memory usage: 24.5 KB


In [ ]:
# Separar as variáveis para o algoritmo
X = df.drop(['mpg'], axis = 1)
y = df['mpg']

In [ ]:
# Normalizar o conjunto das features, pois o algoritmo exige para dados não normalizados
scale = StandardScaler()
X_scale = scale.fit_transform(X)

In [ ]:
# Dividir em dados treino e teste
X_treino, X_teste, y_treino, y_teste = train_test_split(X_scale, y, test_size = 0.3, random_state = 100)

### Avaliação com diferentes configurações dos hiperparametros com GridSearchCV

### Machine Learning com LinearSVR e GridSearchCV

In [ ]:
# Criar objeto LinearSVR, apenas alterando verbose para acompanharmos as info geradas pelo algoritmo.
svr = LinearSVR(random_state=590)

# Criar parâmetros para o LinearSVR com o recurso do Grid (param_grid) em forma de dicionário
param_grid = ({'epsilon': [0.5, 1.0, 1.5],
               'C': [0.001, 0.005, 0.01, 0.05, 0.1, 0.5, 1, 1.5, 2.0, 2.5, 3.0], 
               'loss': ['epsilon_insensitive', 'squared_epsilon_insensitive']})

# Criar objeto GridSearchCV
grid = GridSearchCV(svr,
                    param_grid,
                    cv = 10,
                    scoring = 'r2')

In [ ]:
# treinar o modelo
grid.fit(X_scale, y)

GridSearchCV(cv=10, estimator=LinearSVR(random_state=590),
             param_grid={'C': [0.001, 0.005, 0.01, 0.05, 0.1, 0.5, 1, 1.5, 2.0,
                               2.5, 3.0],
                         'epsilon': [0.5, 1.0, 1.5],
                         'loss': ['epsilon_insensitive',
                                  'squared_epsilon_insensitive']},
             scoring='r2')

## <font color = '#0000FF'>ANÁLISE DOS PARAMETROS</font>

In [ ]:
# Best_estimator: mostra todos os parâmetros com melhor performance apresentado no treinamento
print(grid.best_estimator_)

LinearSVR(C=3.0, epsilon=1.0, loss='squared_epsilon_insensitive',
          random_state=590)


In [ ]:
# best_params_: mostra, dados a configuração do param_grid, as melhores combinações de performance
print(grid.best_params_)

{'C': 3.0, 'epsilon': 1.0, 'loss': 'squared_epsilon_insensitive'}


In [ ]:
# cv_results_: mostra toda a variação de testes realizados dado a configuração do param_grid
print(grid.cv_results_)

{'mean_fit_time': array([0.00100083, 0.00080047, 0.0007004 , 0.00077877, 0.00079982,
       0.00059984, 0.00080659, 0.00070763, 0.00076168, 0.00079951,
       0.00096741, 0.00095255, 0.00080028, 0.00080044, 0.00081303,
       0.00076272, 0.00069952, 0.0009083 , 0.00090775, 0.00060604,
       0.00080726, 0.00090573, 0.0007    , 0.00086019, 0.00076933,
       0.00090714, 0.00080001, 0.00074568, 0.00060081, 0.00084221,
       0.00080009, 0.00203354, 0.00093336, 0.00191152, 0.00079954,
       0.0018389 , 0.00060077, 0.00314054, 0.0008013 , 0.00296445,
       0.00113988, 0.00283203, 0.00123858, 0.00425806, 0.00127618,
       0.00430412, 0.00106847, 0.00360038, 0.00137188, 0.00535169,
       0.0009666 , 0.00513899, 0.00111551, 0.00440407, 0.00137291,
       0.00683632, 0.00136073, 0.00609539, 0.00137081, 0.00557032,
       0.00149987, 0.00797651, 0.00147033, 0.00703938, 0.00118048,
       0.00659428]), 'std_fit_time': array([4.48090385e-04, 4.00234785e-04, 4.58521688e-04, 3.94597344e-04,
   

## <font color = '#0000FF'>ANÁLISE DOS ATRIBUTOS</font>
### Segundo o site do sklearn, atributos são uma forma de como as informações são armazenados em um estimador durante o ajuste. Atributos públicos tem a seguinte estrutura, começam com uma letra e terminam com um sublinhado( _ ), assim 'coef_', 'labels_'.

In [ ]:
# best_score_: mostra o valor da melhor performance dentre todas as combinações, mas representando a combinação 
# vista em best_params
# Neste caso para R2
print((grid.best_score_)*100)

62.01605945257806


In [ ]:
# Extrair o valor do intercept: momento em que x = 0 e cruza o eixo y
# Aqui existe a combinação de dois atributos(.best_estimator e .coef_)
print(grid.best_estimator_.intercept_)

[23.4535391]


In [ ]:
# Extrair os coeficientes de cada variável antes do treinamento
pd.DataFrame(X.columns, grid.best_estimator_.coef_)

0
-0.634462     cylinders
 0.983115  displacement
-0.137565    horsepower
-5.796531        weight
 0.304744  accelaration
 2.829008    model_year

## <font color = '#0000FF'>TREINAMENTO DO MODELO</font> 
### Usaremos os melhores parametros dados pelo .best_params

In [ ]:
# Criar objeto LinearSVR
svr_best = LinearSVR(C = 3, epsilon =  1.0, loss = 'squared_epsilon_insensitive')

In [ ]:
# Treinar modelo e aplicar
modelo = svr_best.fit(X_treino, y_treino)

In [ ]:
# Valor do coeficiente de determinação (R2) após treinamento
modelo.score(X_treino, y_treino)*100

80.91089824635173

In [ ]:
# Ver valor do intercepto. Valor que cruza o eixo y quando x = 0
print(modelo.intercept_)

[23.56126145]


In [ ]:
# Extrair os coeficientes após treinamento
# Valore que explicam a variável target
pd.DataFrame(X.columns, modelo.coef_)

0
-0.341138     cylinders
 1.364088  displacement
-0.727467    horsepower
-5.813947        weight
 0.509077  accelaration
 2.799148    model_year

### Interpretação dos coeficientes
a) quanto mais cilindros, menor a eficiencia do veículo de mpg(milhas por galão)

b) quanto maior o deslocamento do motor, mais o veículo rende por galão

c) horsepower aumentado significa queda de rendimento por galão

d) quanto mais pesado o veículo menor o rendimento por galão

e) quanto maior a aceleração, maior o rendimento por galão

f) quanto mais novo o modelo do veículo maior o rendimento por galão.

#### É estranho a relação proporcional entre displacement e mpg, visto que são grandezas inversas.
#### Não irei resolver esse problema aqui, por não ser o objetivo do post, mas fica pra você analisar com uma gráfico de correlação.
#### Também é possivel eliminar valores outiliers, o que melhoria os valores dos atributos

## <font color = '#228B22'>CONCLUSÕES</font>

### <font color = '#6495ED'>MELHORES PARÂMETROS SELECIONADOS PELO GRIDSEARHCV</font>

LinearSVR:
{'C': 3.0, 'epsilon': 1.0, 'loss': 'squared_epsilon_insensitive'}


### SCORE PARA R2 (%) APÓS O TREINAMENTO
80.91

#### Não era o foco fazer  a previsão neste notebook, mas sim apresentar alguns dos principais parâmetros e atributos do LinearSVR com GridSearchCV.

### Fica o desafio para você fazer as limpezas do dataset e aplicar novamente o LinearSVR com GridSearchCV e ver o quanto melhorará o modelo.

# <font color = '#0000FF'>FIM</font>